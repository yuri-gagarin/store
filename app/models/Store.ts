import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;