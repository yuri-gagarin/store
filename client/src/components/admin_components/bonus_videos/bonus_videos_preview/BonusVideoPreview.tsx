import React from "react";
import { Button, Embed, Item } from "semantic-ui-react";
// css //
import "./css/bonusVideoPreview.css";
// helpers //
import { ConvertDate, setDefaultImage, trimString } from "../../../helpers/displayHelpers";

interface Props {
  bonusVideo: IBonusVideoData;
}

const BonusVideoPreview: React.FC<Props> = ({ bonusVideo }): JSX.Element => {
  const { description, vimeoURL, youTubeURL, createdAt } = bonusVideo;

  if (youTubeURL) {
    return (
      <div className="youtubeBonusVideo">
        <Embed
          url={bonusVideo.youTubeURL}
        />
        <div className="bonusVideoDesc">
          {description}
        </div>
      </div>
    );
  } else if (vimeoURL) {
    return (
      <div>
        <Embed 
          url={bonusVideo.vimeoURL}
        />
        <div className="bonusVideoDesc">
          {description}
        </div>
      </div>
    );
  } else {
    return (
      <div className="bonusVideoErrorDiv">
        <div className="bonusVideoErrorDivMsg">Can't Resolve Video Link</div>
      </div>
    )
  }
};

export default BonusVideoPreview;