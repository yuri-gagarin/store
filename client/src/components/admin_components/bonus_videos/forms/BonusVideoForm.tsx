import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

export type FormState = {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
}

interface Props {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
  handleCreateProduct(data: FormState): void;
  handleUpdateProduct(data: FormState): void;
}


const ProductForm: React.FC<Props> = ({ description, youTubeURL, vimeoURL, handleCreateProduct, handleUpdateProduct }): JSX.Element => {
  
  const [ newForm, setNewForm ] = useState<boolean>(true)
  const [ formState, setFormState ] = useState<FormState>({ description, youTubeURL, vimeoURL });
  const bonusVideoFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  const handleYouTubeURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      youTubeURL: e.target.value
    });
  };
  const handleVimeoURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      vimeoURL: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateProduct(formState);
    } else {
      handleUpdateProduct(formState);
    }
  };

  useEffect(() => {
    if (description || youTubeURL || vimeoURL) {
      setNewForm(false);
    } else {
      setNewForm(true);
    }
  },  [description, youTubeURL, vimeoURL]);

  useEffect(() => {
    if (bonusVideoFormRef.current) {
      const elem = bonusVideoFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [bonusVideoFormRef]);
  
  return (
    <div className="createProductFormHolder" ref={bonusVideoFormRef}>
      <Form id="createProductForm">
        <Form.Field>
          <label>Video YouTube URL:</label>
          <input 
            onChange={handleYouTubeURLChange} 
            placeholder="YouTube URL here ..." 
            value={formState.youTubeURL}
          />
        </Form.Field>
        <Form.Field>
          <label>Video Vimeo URL:</label>
          <input 
            onChange={handleVimeoURLChange} 
            placeholder="Vimeo URL here..." 
            value={formState.vimeoURL}
          />
        </Form.Field>
        <Form.Field
          id='form-textarea-control-opinion'
          control={TextArea}
          label='Video Description'
          onChange={hadnleDescriptionChange}
          placeholder='Video description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button type='submit' onClick={handleSubmit} content= "Save New Bonus Video" />
            : <Button type='submit' onClick={handleSubmit} content= "Update Bonus Video" />
         }  
      </Form>
    </div>
  );
};

export default ProductForm;