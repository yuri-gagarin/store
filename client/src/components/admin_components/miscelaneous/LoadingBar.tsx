import React from "react";
import { Dimmer, Loader, Segment } from "semantic-ui-react";

interface Props {
  message?: string;
}
const LoadingBar: React.FC<Props> = ({ message }): JSX.Element => {

  return (
    <div style={{ height: "50px", width: "100%" }}>
      <Segment>
        <Dimmer active>
          <Loader>{message ? message : "Loading"}</Loader>
        </Dimmer>
      </Segment>
    </div>
  );
};

export default LoadingBar;
