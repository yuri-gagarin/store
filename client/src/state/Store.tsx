import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

