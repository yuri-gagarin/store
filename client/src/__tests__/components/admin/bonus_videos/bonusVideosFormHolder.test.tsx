import React from "react"
// testing utils
import { mount, ReactWrapper, ShallowWrapper } from "enzyme";
// component imports //
import BonusVideoFormHolder from "../../../../components/admin_components/bonus_videos/forms/BonusVideosFormHolder";
import BonusVideoForm from "../../../../components/admin_components/bonus_videos/forms/BonusVideoForm";
import BonusVideoPreview from "../../../../components/admin_components/bonus_videos/bonus_videos_preview/BonusVideoPreview";
import LoadingBar from "../../../../components/admin_components/miscelaneous/LoadingBar";
// state React.Context //
import { IGlobalAppState, TestStateProvider, IGlobalAppContext } from "../../../../state/Store";
// helpers //
import { setMockBonusVideoState } from "../../../../test_helpers/bonusVideoHelpers";

describe("BonusVideoFormHolder Component tests", () => {
  // TEST default Form Holder state //
  describe("Default Form Holder state",  () => {
    let wrapper: ReactWrapper; 

    beforeAll(() => {
      window.scrollTo = jest.fn;
      wrapper = mount(
        <TestStateProvider>
          <BonusVideoFormHolder />
        </TestStateProvider>
      );
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.find(BonusVideoFormHolder)).toMatchSnapshot();
    });
    it("Form Should be closed by default", () => {
      const form = wrapper.find(BonusVideoForm);
      expect(form.length).toEqual(0);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find("#bonusVideoFormToggleBtn");
      expect(toggleButton.length).toEqual(1);
    });
    it("Should NOT render'#bonusVideoFormDetails' for non-existing BonusVideo data", () => {
      expect(wrapper.find("#bonusVideoFormDetails").length).toEqual(0);
    });
    it("Should NOT render 'BonusVideoPreview' Component", () => {
      expect(wrapper.find(BonusVideoPreview).length).toEqual(0);
    });
  });
  // END default Form Holder State tests //
  // TEST default Form Holder state - Form OPEN - NO BonusVideo data//
  describe("Default Form Holder state - Form Open",  () => {
    let wrapper: ReactWrapper;
    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount(
        <TestStateProvider>
          <BonusVideoFormHolder />
        </TestStateProvider>
      );
      wrapper.find(BonusVideoFormHolder).find("#bonusVideoFormToggleBtn").at(0).simulate("click");
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.find(BonusVideoFormHolder)).toMatchSnapshot();
    });
    it("Form Should be open", () => {
      const form = wrapper.find(BonusVideoForm);
      expect(form.length).toEqual(1);
    });
    
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.find(BonusVideoFormHolder).render().find("#bonusVideoFormToggleBtn");
      expect(toggleButton.length).toEqual(1);
    });
    it("Should NOT render'#bonusVideoFormDetails' for non-existing BonusVideo data", () => {
      const formDetails = wrapper.find(BonusVideoFormHolder).render().find("#bonusVideoFormDetails");
      expect(formDetails.length).toEqual(0);
    });
  
    it("Should NOT render 'BonusVideoPreview' Component", () => {
      const bonusVideoPreview = wrapper.find(BonusVideoPreview);
      expect(bonusVideoPreview.length).toEqual(0);
    });
    it("Should render a 'BonusVide0Form' Component", () => {
      expect(wrapper.find(BonusVideoForm).length).toEqual(1);
    });
  });
  // END default FormHolder state - Form OPEN - NO BonusVideo DATA //
  // TEST Form Holder state OPEN - WITH Current BonusVideo Data //
  describe("Form Holder state OPEN - WITH Current BonusVideo Data",  () => {
    let wrapper: ReactWrapper; let state: IGlobalAppState;

    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockBonusVideoState({ currentBonusVideo: true });
      wrapper = mount(
        <TestStateProvider mockState={state}>
          <BonusVideoFormHolder />
        </TestStateProvider>
      );
      wrapper.find(BonusVideoFormHolder).find("#bonusVideoFormToggleBtn").at(0).simulate("click");
    });

    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#bonusVideoFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should render 'BonusVideoPreview' Component", () => {
      const bonusVideoPreview = wrapper.find(BonusVideoPreview);
      expect(bonusVideoPreview.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const form = wrapper.find(BonusVideoForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.find(BonusVideoFormHolder).render().find("#bonusVideoFormUpdate");
      expect(toggleButton.length).toEqual(1);
    });
    it("Should render'#bonusVideoFormDetails'", () => {
      const bonusVideFormDetails = wrapper.find(BonusVideoFormHolder).render().find("#bonusVideoFormDetails");
      expect(bonusVideFormDetails.length).toEqual(1);
    });
    it("Should correctly render '.bonusVideoFormHolderYouTubeUrl'", () => {
      const videoDetail = wrapper.find(BonusVideoFormHolder).render().find(".bonusVideoFormHolderYouTubeUrl");
      expect(videoDetail.length).toEqual(1);
    });
    it("Should correctly render '.bonusVideoFormHolderVimeoUrl'", () => {
      const videoDetail = wrapper.find(BonusVideoFormHolder).render().find(".bonusVideoFormHolderVimeoUrl");
      expect(videoDetail.length).toEqual(1);
    });
    it("Should correctly render '.bonusVideoFormHolderDescription'", () => {
      const videoDetail = wrapper.find(BonusVideoFormHolder).render().find(".bonusVideoFormHolderDescription");
      expect(videoDetail.length).toEqual(1);
    });
    it("Should correctly render '.bonusVideoFormHolderTimestamps'", () => {
      const videoDetail = wrapper.find(BonusVideoFormHolder).render().find(".bonusVideoFormHolderTimestamps");
      expect(videoDetail.length).toEqual(1);
    });
  });
  // END Form Holder state OPEN - WITH Current BonusVideo Data //
});