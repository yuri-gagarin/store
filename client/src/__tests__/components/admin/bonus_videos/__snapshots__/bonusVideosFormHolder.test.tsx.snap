// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BonusVideoFormHolder Component tests Default Form Holder state - Form Open Should Properly render FormHolder 1`] = `
<BonusVideoFormHolder>
  <div
    id="bonusVideoFormHolder"
  >
    <GridRow>
      <div
        className="row"
      >
        <GridColumn
          computer={14}
          mobile={16}
          tablet={15}
        >
          <div
            className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
          >
            <Button
              as="button"
              content="Close Form"
              id="bonusVideoFormToggleBtn"
              onClick={[Function]}
            >
              <Ref
                innerRef={
                  Object {
                    "current": <button
                      class="ui button"
                      id="bonusVideoFormToggleBtn"
                    >
                      Close Form
                    </button>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="bonusVideoFormToggleBtn"
                      >
                        Close Form
                      </button>,
                    }
                  }
                >
                  <button
                    className="ui button"
                    id="bonusVideoFormToggleBtn"
                    onClick={[Function]}
                  >
                    Close Form
                  </button>
                </RefFindNode>
              </Ref>
            </Button>
            <BonusVideoForm
              description=""
              handleCreateBonusVideo={[Function]}
              handleUpdateBonusVideo={[Function]}
              newForm={true}
              vimeoURL=""
              youTubeURL=""
            >
              <div
                id="createBonusVideoForm"
              >
                <Form
                  as="form"
                >
                  <form
                    className="ui form"
                    onSubmit={[Function]}
                  >
                    <FormField>
                      <div
                        className="field"
                      >
                        <label>
                          Video YouTube URL:
                        </label>
                        <input
                          id="bonusVideoYoutubeInput"
                          onChange={[Function]}
                          placeholder="YouTube URL here ..."
                          value=""
                        />
                      </div>
                    </FormField>
                    <FormField>
                      <div
                        className="field"
                      >
                        <label>
                          Video Vimeo URL:
                        </label>
                        <input
                          id="bonusVideoVimeoInput"
                          onChange={[Function]}
                          placeholder="Vimeo URL here..."
                          value=""
                        />
                      </div>
                    </FormField>
                    <FormField
                      control={[Function]}
                      id="bonusVideoDescInput"
                      label="Video Description"
                      onChange={[Function]}
                      placeholder="Video description here..."
                      value=""
                    >
                      <div
                        className="field"
                      >
                        <label
                          htmlFor="bonusVideoDescInput"
                        >
                          Video Description
                        </label>
                        <TextArea
                          aria-describedby={null}
                          as="textarea"
                          id="bonusVideoDescInput"
                          onChange={[Function]}
                          placeholder="Video description here..."
                          rows={3}
                          value=""
                        >
                          <Ref
                            innerRef={
                              Object {
                                "current": <textarea
                                  id="bonusVideoDescInput"
                                  placeholder="Video description here..."
                                  rows="3"
                                />,
                              }
                            }
                          >
                            <RefFindNode
                              innerRef={
                                Object {
                                  "current": <textarea
                                    id="bonusVideoDescInput"
                                    placeholder="Video description here..."
                                    rows="3"
                                  />,
                                }
                              }
                            >
                              <textarea
                                aria-describedby={null}
                                id="bonusVideoDescInput"
                                onChange={[Function]}
                                onInput={[Function]}
                                placeholder="Video description here..."
                                rows={3}
                                value=""
                              />
                            </RefFindNode>
                          </Ref>
                        </TextArea>
                      </div>
                    </FormField>
                    <Button
                      as="button"
                      content="Save New Bonus Video"
                      id="bonusVideoFormCreate"
                      onClick={[Function]}
                      type="submit"
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": <button
                              class="ui button"
                              id="bonusVideoFormCreate"
                              type="submit"
                            >
                              Save New Bonus Video
                            </button>,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": <button
                                class="ui button"
                                id="bonusVideoFormCreate"
                                type="submit"
                              >
                                Save New Bonus Video
                              </button>,
                            }
                          }
                        >
                          <button
                            className="ui button"
                            id="bonusVideoFormCreate"
                            onClick={[Function]}
                            type="submit"
                          >
                            Save New Bonus Video
                          </button>
                        </RefFindNode>
                      </Ref>
                    </Button>
                  </form>
                </Form>
              </div>
            </BonusVideoForm>
          </div>
        </GridColumn>
      </div>
    </GridRow>
  </div>
</BonusVideoFormHolder>
`;

exports[`BonusVideoFormHolder Component tests Default Form Holder state Should Properly Mount Form Holder 1`] = `
<BonusVideoFormHolder>
  <div
    id="bonusVideoFormHolder"
  >
    <GridRow>
      <div
        className="row"
      >
        <GridColumn
          computer={14}
          mobile={16}
          tablet={15}
        >
          <div
            className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
          >
            <Button
              as="button"
              content="Open Form"
              id="bonusVideoFormToggleBtn"
              onClick={[Function]}
            >
              <Ref
                innerRef={
                  Object {
                    "current": <button
                      class="ui button"
                      id="bonusVideoFormToggleBtn"
                    >
                      Open Form
                    </button>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="bonusVideoFormToggleBtn"
                      >
                        Open Form
                      </button>,
                    }
                  }
                >
                  <button
                    className="ui button"
                    id="bonusVideoFormToggleBtn"
                    onClick={[Function]}
                  >
                    Open Form
                  </button>
                </RefFindNode>
              </Ref>
            </Button>
          </div>
        </GridColumn>
      </div>
    </GridRow>
  </div>
</BonusVideoFormHolder>
`;

exports[`BonusVideoFormHolder Component tests Form Holder state OPEN - WITH Current BonusVideo Data Should properly render BonusVideoFormHolder component 1`] = `
<BonusVideoFormHolder>
  <div
    id="bonusVideoFormHolder"
  >
    <GridRow
      id="bonusVideoFormDetails"
    >
      <div
        className="row"
        id="bonusVideoFormDetails"
      >
        <GridColumn
          computer={14}
          mobile={16}
          tablet={14}
        >
          <div
            className="fourteen wide computer sixteen wide mobile fourteen wide tablet column"
          >
            <h1>
              Details
            </h1>
            <div
              className="bonusVideoFormHolderYouTubeUrl"
            >
              <h3>
                Video youTube URL:
              </h3>
              <p>
                url
              </p>
            </div>
            <div
              className="bonusVideoFormHolderVimeoUrl"
            >
              <h3>
                Video vimeo URL:
              </h3>
              <p>
                url
              </p>
            </div>
            <div
              className="bonusVideoFormHolderDescription"
            >
              <h3>
                Video description:
              </h3>
              <p>
                description
              </p>
            </div>
            <div
              className="bonusVideoFormHolderTimestamps"
            >
              <span>
                Created At: 
                <strong>
                  NaN-undefined-NaN
                </strong>
              </span>
              <span>
                Edited At: 
                <strong>
                  Not Available
                </strong>
              </span>
            </div>
            <div
              className="bonusVideoFormPreview"
            >
              <BonusVideoPreview
                bonusVideo={
                  Object {
                    "_id": "1",
                    "createdAt": "now",
                    "description": "description",
                    "vimeoURL": "url",
                    "youTubeURL": "url",
                  }
                }
              >
                <div
                  className="youtubeBonusVideo"
                >
                  <Embed
                    url="url"
                  >
                    <div
                      className="ui embed"
                      onClick={[Function]}
                    >
                      <Icon
                        as="i"
                        name="video play"
                      >
                        <i
                          aria-hidden="true"
                          className="video play icon"
                          onClick={[Function]}
                        />
                      </Icon>
                    </div>
                  </Embed>
                  <div
                    className="bonusVideoDesc"
                  >
                    description
                  </div>
                </div>
              </BonusVideoPreview>
            </div>
          </div>
        </GridColumn>
      </div>
    </GridRow>
    <GridRow>
      <div
        className="row"
      >
        <GridColumn
          computer={14}
          mobile={16}
          tablet={15}
        >
          <div
            className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
          >
            <Button
              as="button"
              content="Close Form"
              id="bonusVideoFormToggleBtn"
              onClick={[Function]}
            >
              <Ref
                innerRef={
                  Object {
                    "current": <button
                      class="ui button"
                      id="bonusVideoFormToggleBtn"
                    >
                      Close Form
                    </button>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="bonusVideoFormToggleBtn"
                      >
                        Close Form
                      </button>,
                    }
                  }
                >
                  <button
                    className="ui button"
                    id="bonusVideoFormToggleBtn"
                    onClick={[Function]}
                  >
                    Close Form
                  </button>
                </RefFindNode>
              </Ref>
            </Button>
            <BonusVideoForm
              description="description"
              handleCreateBonusVideo={[Function]}
              handleUpdateBonusVideo={[Function]}
              newForm={false}
              vimeoURL="url"
              youTubeURL="url"
            >
              <div
                id="createBonusVideoForm"
              >
                <Form
                  as="form"
                >
                  <form
                    className="ui form"
                    onSubmit={[Function]}
                  >
                    <FormField>
                      <div
                        className="field"
                      >
                        <label>
                          Video YouTube URL:
                        </label>
                        <input
                          id="bonusVideoYoutubeInput"
                          onChange={[Function]}
                          placeholder="YouTube URL here ..."
                          value="url"
                        />
                      </div>
                    </FormField>
                    <FormField>
                      <div
                        className="field"
                      >
                        <label>
                          Video Vimeo URL:
                        </label>
                        <input
                          id="bonusVideoVimeoInput"
                          onChange={[Function]}
                          placeholder="Vimeo URL here..."
                          value="url"
                        />
                      </div>
                    </FormField>
                    <FormField
                      control={[Function]}
                      id="bonusVideoDescInput"
                      label="Video Description"
                      onChange={[Function]}
                      placeholder="Video description here..."
                      value="description"
                    >
                      <div
                        className="field"
                      >
                        <label
                          htmlFor="bonusVideoDescInput"
                        >
                          Video Description
                        </label>
                        <TextArea
                          aria-describedby={null}
                          as="textarea"
                          id="bonusVideoDescInput"
                          onChange={[Function]}
                          placeholder="Video description here..."
                          rows={3}
                          value="description"
                        >
                          <Ref
                            innerRef={
                              Object {
                                "current": <textarea
                                  id="bonusVideoDescInput"
                                  placeholder="Video description here..."
                                  rows="3"
                                >
                                  description
                                </textarea>,
                              }
                            }
                          >
                            <RefFindNode
                              innerRef={
                                Object {
                                  "current": <textarea
                                    id="bonusVideoDescInput"
                                    placeholder="Video description here..."
                                    rows="3"
                                  >
                                    description
                                  </textarea>,
                                }
                              }
                            >
                              <textarea
                                aria-describedby={null}
                                id="bonusVideoDescInput"
                                onChange={[Function]}
                                onInput={[Function]}
                                placeholder="Video description here..."
                                rows={3}
                                value="description"
                              />
                            </RefFindNode>
                          </Ref>
                        </TextArea>
                      </div>
                    </FormField>
                    <Button
                      as="button"
                      content="Update Bonus Video"
                      id="bonusVideoFormUpdate"
                      onClick={[Function]}
                      type="submit"
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": <button
                              class="ui button"
                              id="bonusVideoFormUpdate"
                              type="submit"
                            >
                              Update Bonus Video
                            </button>,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": <button
                                class="ui button"
                                id="bonusVideoFormUpdate"
                                type="submit"
                              >
                                Update Bonus Video
                              </button>,
                            }
                          }
                        >
                          <button
                            className="ui button"
                            id="bonusVideoFormUpdate"
                            onClick={[Function]}
                            type="submit"
                          >
                            Update Bonus Video
                          </button>
                        </RefFindNode>
                      </Ref>
                    </Button>
                  </form>
                </Form>
              </div>
            </BonusVideoForm>
          </div>
        </GridColumn>
      </div>
    </GridRow>
  </div>
</BonusVideoFormHolder>
`;
